resilience4j-retry/src/main/java/io/github/resilience4j/retry/internal/RetryBudgetImpl.java

```diff
@@ -0,0 +1,255 @@
+                throwOrSleepAfterException();
+            } else {
+                failedWithoutRetryCounter.increment();
+                publishRetryEvent(() -> new RetryOnIgnoredErrorEvent(getName(), exception));
+                throw exception;
+            }
+        }
+
+        public void onRuntimeError(RuntimeException runtimeException) {
+            if (exceptionPredicate.test(runtimeException)) {
+                lastRuntimeException.set(runtimeException);
+                throwOrSleepAfterRuntimeException();
+            } else {
+                failedWithoutRetryCounter.increment();
+                publishRetryEvent(() -> new RetryOnIgnoredErrorEvent(getName(), runtimeException));
+                throw runtimeException;
+            }
+        }
+
+        private void throwOrSleepAfterException() throws Exception {
+            int currentNumOfAttempts = numOfAttempts.incrementAndGet();
+            if (canRetry(currentNumOfAttempts)) {
+                waitIntervalAfterFailure();
+            } else {
+                failedAfterRetryCounter.increment();
+                Exception throwable = lastException.get();
+                publishRetryEvent(() -> new RetryOnErrorEvent(getName(), currentNumOfAttempts, throwable));
+                throw throwable;
+            }
+        }
+
+        private void throwOrSleepAfterRuntimeException() {
+            int currentNumOfAttempts = numOfAttempts.incrementAndGet();
+            if (canRetry(currentNumOfAttempts)) {
+                waitIntervalAfterFailure();
+            } else {
+                failedAfterRetryCounter.increment();
+                RuntimeException throwable = lastRuntimeException.get();
+                publishRetryEvent(() -> new RetryOnErrorEvent(getName(), currentNumOfAttempts, throwable));
+                throw throwable;
+            }
+        }
+
+        private void waitIntervalAfterFailure() {
+            // wait interval until the next attempt should start
+            long interval = intervalFunction.apply(numOfAttempts.get());
+            Try.run(() -> sleepFunction.accept(interval))
+                    .getOrElseThrow(ex -> lastRuntimeException.get());
+        }
+
+        /**
+         * Resolves whether the failed operation can retry.
+         *
+         * @param currentNumOfAttempts current number of attempts by this operation
+         * @return boolean whether the operation can retry
+         */
+        private boolean canRetry(final int currentNumOfAttempts) {
+            // check max number of attempts
+            if (currentNumOfAttempts >= maxAttempts) {
+                // mark next as '0'
+                ringBitSet.setNextBit(false);
+                return false;
+            }
+            synchronized (this) {
+                // retrieve number of retries in bit set, add one and divided by buffer size
+                double retriesPctWithOneMore = (ringBitSet.cardinality() + 1) / bufferSize;
```

Comment:
I think you should call `ringBitSet.setNextBit` only from `onSuccess` and `onError` methods. And `canRetry` should only check threshold without any additional synchronisations. Perhaps, I'm missing something because of trying to read completely from GitHub, so correct me if I'm wrong.